server/datastore/mysql/migrations/tables/20240221112844_FixUniquePolicyNameBug.go

```diff
@@ -0,0 +1,107 @@
+package tables
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"github.com/VividCortex/mysqlerr"
+	"github.com/go-sql-driver/mysql"
+)
+
+func init() {
+	MigrationClient.AddMigration(Up_20240221112844, Down_20240221112844)
+}
+
+func Up_20240221112844(tx *sql.Tx) error {
+	// The bug was that the checksum was not updating. So we need to update the checksum for existing rows.
+	const updateStmt = `
+	UPDATE
+		policies
+	SET
+		checksum = UNHEX(
+			MD5(
+				-- concatenate with separator \x00
+				CONCAT_WS(CHAR(0),
+					COALESCE(team_id, ''),
+					name
+				)
+			)
+		)
+	`
+	const updateNameStmt = `
+	UPDATE
+		policies
+	SET
+		name = CONCAT(name, '%d'),
+		checksum = UNHEX(
+			MD5(
+				-- concatenate with separator \x00
+				CONCAT_WS(CHAR(0),
+					COALESCE(team_id, ''),
+					name
+				)
+			)
+		)
+	WHERE id = ?
+	`
+	_, err := tx.Exec(updateStmt)
+	isDuplicate := func(err error) bool {
+		var driverErr *mysql.MySQLError
+		if errors.As(err, &driverErr) && driverErr.Number == mysqlerr.ER_DUP_ENTRY {
+			return true
+		}
+		return false
+	}
+
+	duplicateExists := false
+	if err != nil {
+		if isDuplicate(err) {
+			duplicateExists = true
+		} else {
+			return fmt.Errorf("failed to update policies table to fill the checksum column: %w", err)
+		}
+	}
+
+	// Since there is a duplicate name somewhere in the table (should be rare), we update table 1 row at a time
+	if duplicateExists {
+		rows, err := tx.Query("SELECT id FROM policies")
+		if err != nil {
+			return fmt.Errorf("failed to query policies table: %w", err)
+		} else if rows.Err() != nil {
+			return fmt.Errorf("failed to query policies table: %w", rows.Err())
+		}
+		var ids []uint64
+		{
+			defer rows.Close()
+			for rows.Next() {
+				var id uint64
+				if err := rows.Scan(&id); err != nil {
+					return fmt.Errorf("failed to scan policies table: %w", err)
+				}
+				ids = append(ids, id)
+			}
+		}
+		for _, id := range ids {
+			_, err = tx.Exec(updateStmt+" WHERE id = ?", id)
+			if err != nil && isDuplicate(err) {
+				for i := 2; i < 10000; i++ {
+					_, err = tx.Exec(fmt.Sprintf(updateNameStmt, i), id)
+					if err != nil && isDuplicate(err) {
+						continue
+					}
+					// We do not update this row -- it can be updated next time the policy is modified. This should be very rare.
+					// Lack of update can happen if duplicate name is 255 characters, or all of the nearly 10000 names we tried are already taken.
+					break
```

Comment:
Will users have to update their yamls? (to append the ` N` on the policy names).

Reply:
If users have 2 policies with the same name, their yaml will apply to the one that was originally created with that name. I don't expect users to actually have policies with the same names.

Reply:
> If users have 2 policies with the same name, their yaml will apply to the one that was originally created with that name.

Gotcha. Also maybe we are a-ok because of this check?
https://github.com/fleetdm/fleet/blob/23223d3793b73174ba123504fd5a5cc6ceaa0600/server/service/global_policies.go#L522

> I don't expect users to actually have policies with the same names.

Didn't we hit this in dogfood?